### DISCUSSION TITLE
TimescaleDB sink: Conditionally insert null into column based on metric content

### DISCUSSION THREAD

**Discussion by mikolajwalus** [emqx/emqx#14109] [Q&A] [answered]
Hi everyone,

I have question regarding rule engine. I'm trying to configure enterprise version broker to be able to insert data into timescale.

I want to conditionally insert null into into column based on metric content

Unfortunately I have following problem I'm unable to resolve.

Let's say this is my table:

```
CREATE TABLE sensor_data (
    ts                      TIMESTAMPTZ       NOT NULL,
    metric_name    TEXT                       NOT NULL,
    device_id          TEXT                       NOT NULL,
    value_int16       INT                         NULL,
    value_bool        BOOLEAN              NULL
);
```

I'm doing this as a query to extract all the values

```
FOREACH
    payload.metrics as metric
DO
    metric.n as metric_name,
    metric.v as v,
    unix_ts_to_rfc3339(metric.ts, 'millisecond') as ts,
    CASE WHEN metric.dt = 6 THEN metric.v
      ELSE ''
    END as value_int16,
    CASE WHEN metric.dt = 7 THEN metric.v
      ELSE ''
    END as value_bool,
    substr(last(split(topic, '/')), 2) as device_id
INCASE
  regex_match(metric.n, '^event.*') = true
FROM "metrics/#"
```

problem is that I want to store a metric in the column based of the data type (dt in metric)). Unfortunately i get error that indicates that I try to insert empty string into boolean table.

I tried also this:

```
FOREACH
    payload.metrics as metric
DO
    metric.n as metric_name,
    unix_ts_to_rfc3339(metric.ts, 'millisecond') as ts,
    CASE WHEN metric.dt = 6 THEN metric.v
      ELSE null
    END as value_int16,
    CASE WHEN metric.dt = 7 THEN metric.v
      ELSE null
    END as value_bool,
    substr(last(split(topic, '/')), 2) as device_id
INCASE
  regex_match(metric.n, '^event.*') = true
FROM "metrics/#"
```

but here I get similar error with "undefined".

Here is my insert statement:

```
INSERT INTO
 sensor_data (ts, metric_name, device_id, value_int16, value_bool)
VALUES
   (${ts}, ${metric_name}, ${device_id}, ${value_int16}, ${value_bool})
```

Is there any workaround for this? I saw some "Undefined Vars as Null switch" in MySql connector but nothing like that for TimescaleDB or Postgres.

I was thinking maybe about doing something like this

```
INSERT INTO
 sensor_data (ts, metric_name, device_id, value_int16, value_bool)
VALUES
   (${ts}, ${metric_name}, ${device_id}, ${value_int16}, NULLIF(${value_bool}, '')::boolean)
```

but it looks like dirty hack for me

**Comment by zmstone** [comment]
```
    CASE WHEN metric.dt = 7 THEN metric.v
      ELSE null
    END as value_bool,
```
are you sure `metric.v` is not empty sting?

**Reply by zmstone** [reply]
there is a `coalesce_ne` function which means coalesce when not empty, it returns the second arg if the first is undefined or empty. 
so maybe try this:
```
    CASE WHEN metric.dt = 7 THEN coalesce_ne(metric.v, null)
      ELSE null
    END as value_bool,
```

**Reply by mikolajwalus** [reply]
Hi,

Thank you for response.

When i change rule to this: 
```
FOREACH
    payload.metrics as metric
DO
    metric.n as metric_name,
    metric.v as v,
    unix_ts_to_rfc3339(metric.ts, 'millisecond') as ts,
    CASE WHEN metric.dt = 6 THEN metric.v
      ELSE ''
    END as value_int16,
    CASE WHEN metric.dt = 7 THEN coalesce_ne(metric.v, null)
      ELSE null
    END as value_bool,
    substr(last(split(topic, '/')), 2) as device_id
INCASE
  regex_match(metric.n, '^event.*') = true
FROM "secomea/#"
```

and have following insert template:
```
INSERT INTO
 sensor_data (ts, metric_name, device_id, value_int16, value_bool)
VALUES
   (${ts}, ${metric_name}, ${device_id}, ${value_int16}, ${value_bool})
```

I got following error in logs:

> [error] Insert failed reason: {error,error,<<"22P02">>,invalid_text_representation,<<"invalid input syntax for type boolean: \"undefined\"">>,[{file,<<"bool.c">>},{line,<<"151">>},{position,<<"189">>},{routine,<<"boolin">>},{severity,<<"ERROR">>}]}


Looks like null from rule engine is treated as "undefined" (that is exactly what I can see when I test the rule). This "undefined" is treated as string is tried to `be inserted into boolean column (what error says) but I just wanted to insert null because there is no data.

**Reply by zmstone** [reply]
I’m afk, could you try to quote null? ‘null’

**Reply by mikolajwalus** [reply]
With this (ELSE 'null'):
```
    CASE WHEN metric.dt = 7 THEN coalesce_ne(metric.v, null)
      ELSE 'null'
    END as value_bool,
```

I have: 

> invalid input syntax for type boolean: \"null\"

With this (coalesce_ne(metric.v, 'null')):
```
    CASE WHEN metric.dt = 7 THEN coalesce_ne(metric.v, 'null')
      ELSE null
    END as value_bool,
```

I have: 

> "invalid input syntax for type boolean: \"undefined\""

**Reply by mikolajwalus** [reply]
And I want to point out that this part work fine if dt = 6.

```
CASE WHEN metric.dt = 6 THEN metric.v
      ELSE ''
    END as value_int16,
```

So metric.v is an ordinary number

**Reply by zmstone** [reply]
Is metric.v a string value “undefined”?

**Reply by mikolajwalus** [reply]
No metric.v is 10 and if I remove `${value_bool}` from insert like below works fine.
```
INSERT INTO
 sensor_data (ts, metric_name, device_id, value_int16, value_bool)
VALUES
   (${ts}, ${metric_name}, ${device_id}, ${value_int16}, NULL)
 ```
 
 But I cannot set value_bool to null on rule engine level so it will be inserted properly to db.

**Reply by zmstone** [reply]
if metric.v is 10, how do you expect it to be a boolean ?

```
    CASE WHEN metric.dt = 7 THEN metric.v
      ELSE ''
    END as value_bool,
```
this means if metric.dt is 7, then let  `value_bool` be `metric.v` which is supposed to be a boolean,

**Reply by mikolajwalus** [reply]
If value is 10 than metric.dt is not 7 because sent value is probably a boolean and in thios case I want value_bool be inserted as null. In your latest reply you made a behavior that I would like to have but I wasn't able to reproduce it. I asked there about insert template maybe there is something I'm missing.

**Comment by zmstone** [comment]
I tied to test with simplified data set with postgres.

```
FOREACH
    payload.metrics as metric
DO
    clientid as clientid,
    metric.device_id as device_id,
    CASE WHEN metric.dt = 7 THEN coalesce_ne(metric.v, null)
      ELSE null
    END as value_bool
FROM "metrics/#"
```

send data like this:

```
#!/bin/bash

payload='
    {
        "metrics":
            [
                {"dt": 6,
                 "device_id": "1"
                },
                {"dt": 7,
                 "device_id": "2"
                },
                {"dt": 7,
                 "v": false,
                 "device_id": "3"
                },
                {"dt": 7,
                 "v": true,
                 "device_id": "4"
                }
            ]
    }
'

mqttx pub -h localhost -t metrics/a -m "${payload}"
```

and query from postgres:

```
   client_id    | devid | boolval
----------------+-------+---------
 mqttx_89ed2e4b | 1     |
 mqttx_89ed2e4b | 2     |
 mqttx_89ed2e4b | 3     | f
 mqttx_89ed2e4b | 4     | t
```

and the rule is like this;

**Reply by mikolajwalus** [reply]
Looks exactly how I wanted it to work but how looks your data insert template?

Maybe there is something I'm missing,

**Reply by zmstone** [reply, marked as answer]
there is nothing special about the template. 
Here are the possible relevant differences
- I tested against postgres, not timescale
- I tested with EMQX Enterprise 5.8.1 (not sure which version you are using?)
- I have `disable_prepared_statements = false` in the connector config

**Reply by zmstone** [reply]
I remembered that there was a bug before 5.8.1: after you have updated the SQL insert template, if the connection towards postgres/timescale had to restart, the old prepared statement (before your update) may take effect.
this has been fixed in 5.8.1 in this pr: https://github.com/emqx/emqx/pull/13901

**Reply by mikolajwalus** [reply]
I'm right now on vacations but I'll check out what you highlighted immediately on monday or sunday( if'll have access to my PC)

But to clarify this is would be your insert statement? (no nullif ect)

```
INSERT INTO
 sensor_data (client_id, devid, boolval)
VALUES
   (${clientid}, ${device_id}, ${value_bool})
 ```

**Reply by zmstone** [reply]
yes, that's my template, no nullif.

**Reply by mikolajwalus** [reply]
Hi, I was using old version of EMQX broker. After updating to 5.8.1 nulls are handled perfectly fine. 

Really thank you for your support
